File: sdks/ruby/examples/transform_data_node.rb
# frozen_string_literal: true

require_relative "../lib/blok"

# TransformDataNode transforms JSON data based on field mappings.
#
# Config:
#   - "mappings"     (hash, optional): Map of target field -> source field path (dot-notation)
#   - "include_only" (array, optional): Only include these fields in output
#   - "exclude"      (array, optional): Exclude these fields from output
#   - "defaults"     (hash, optional): Default values for missing fields
#
# Input: Request body (must be a JSON object)
# Output: Transformed JSON object
class TransformDataNode < Blok::Node::NodeHandler
  def execute(ctx, config)
    body = ctx.request.body
    unless body.is_a?(Hash)
      raise Blok::Errors::NodeError.validation("request body must be a JSON object")
    end

    result = {}

    # Apply field mappings if configured
    if config["mappings"].is_a?(Hash)
      config["mappings"].each do |target, source_path|
        next unless source_path.is_a?(String)

        value = get_nested_value(body, source_path)
        result[target] = value unless value.nil?
      end
    else
      # No mappings -- copy all fields
      body.each { |k, v| result[k] = v }
    end

    # Apply include_only filter
    if config["include_only"].is_a?(Array)
      allowed = config["include_only"]
      result.select! { |k, _| allowed.include?(k) }
    end

    # Apply exclude filter
    if config["exclude"].is_a?(Array)
      config["exclude"].each { |field| result.delete(field) }
    end

    # Apply defaults for missing fields
    if config["defaults"].is_a?(Hash)
      config["defaults"].each do |k, v|
        result[k] = v unless result.key?(k)
      end
    end

    # Store in context for downstream nodes
    ctx.set_var("transformed_data", result)

    result
  end

  private

  # Traverse dot-notation path (e.g. "user.name") into nested hashes.
  def get_nested_value(data, path)
    current = data
    path.split(".").each do |part|
      return nil unless current.is_a?(Hash)

      current = current[part]
    end
    current
  end
end

# ----- Boot the server if run directly -----
if __FILE__ == $PROGRAM_NAME
  registry = Blok::Server::RuntimeApp.registry
  registry.register("transform-data", TransformDataNode.new)

  puts "Starting TransformDataNode on port 8080..."
  Blok::Server::RuntimeApp.run!(port: 8080, bind: "0.0.0.0")
end

File: sdks/python3/examples/transform_data_node.py
from __future__ import annotations
from typing import Any, Dict

from blok.node.node_handler import NodeHandler
from blok.types.context import Context
from blok.errors.node_error import NodeError


class TransformDataNode(NodeHandler):
    """Transforms JSON data based on field mappings.

    Config:
        mappings (dict, optional): Map of target field -> source field path (dot-notation)
        include_only (list, optional): Only include these fields in output
        exclude (list, optional): Exclude these fields from output
        defaults (dict, optional): Default values for missing fields

    Input: Request body (must be a JSON object)
    Output: Transformed JSON object
    """

    def execute(self, ctx: Context, config: Dict[str, Any]) -> Any:
        body = ctx.request.body_map()
        if body is None:
            raise NodeError.validation("request body must be a JSON object")

        result: Dict[str, Any] = {}

        # Apply field mappings if configured
        mappings = config.get("mappings")
        if isinstance(mappings, dict):
            for target_field, source_path in mappings.items():
                if not isinstance(source_path, str):
                    continue
                value = _get_nested_value(body, source_path)
                if value is not None:
                    result[target_field] = value
        else:
            # No mappings -- copy all fields
            result = dict(body)

        # Apply include_only filter
        include_only = config.get("include_only")
        if isinstance(include_only, list) and include_only:
            result = {k: v for k, v in result.items() if k in include_only}

        # Apply exclude filter
        exclude = config.get("exclude")
        if isinstance(exclude, list):
            for field in exclude:
                result.pop(field, None)

        # Apply defaults for missing fields
        defaults = config.get("defaults")
        if isinstance(defaults, dict):
            for k, v in defaults.items():
                if k not in result:
                    result[k] = v

        ctx.set_var("transformed_data", result)
        return result


def _get_nested_value(data: Dict[str, Any], path: str) -> Any:
    """Traverse dot-notation path into nested dicts."""
    current: Any = data
    for part in path.split("."):
        if not isinstance(current, dict):
            return None
        current = current.get(part)
    return current
